### PR TITLE
support adding to usergroups directly

### PR DESCRIPTION
**Pull request by pjain1**
Fixes https://github.com/rilldata/rill/issues/5349

- For non existing Rill user, invite is sent and upon login user gets added to org as viewer and added to desired usergroup
- For existing Rill user, confirmation is asked to add user to org and upon confirmation, gets added to the desired usergroup

### CODE REVIEW

File: admin/server/usergroups.go [hunk 1]
```diff
@@ -487,22 +488,99 @@ func (s *Server) AddUsergroupMemberUser(ctx context.Context, req *adminv1.AddUse
...
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+
+			err = s.admin.DB.InsertUsergroupMemberUser(ctx, *org.AllUsergroupID, user.ID)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+		} else {
+			return nil, status.Error(codes.FailedPrecondition, "user is not a member of the organization")
+		}
```
Comment (begelundmuller) [inline admin/server/usergroups.go:583]: You can reduce nesting by changing the case to `if !req.AddRillUserToOrg { ... }`. See the style guide: https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting

File: cli/cmd/user/add.go [hunk 2]
```diff
@@ -37,16 +39,44 @@ func AddCmd(ch *cmdutil.Helper) *cobra.Command {
...
+				res, err := client.AddUsergroupMemberUser(cmd.Context(), &adminv1.AddUsergroupMemberUserRequest{
+					Organization:     ch.Org,
+					Usergroup:        group,
+					Email:            email,
+					AddRillUserToOrg: false,
 				})
 				if err != nil {
+					if s, ok := status.FromError(err); ok && s.Code() == codes.FailedPrecondition && strings.Contains(s.Message(), "user is not a member of the org") {
+						// ask if user should be added to the organization
+						addToOrg, err := cmdutil.ConfirmPrompt(fmt.Sprintf("This user exists in Rill but not part of the %q organization. Do you want to add the user to the organization as viewer?", ch.Org), "", false)
```
Comment (begelundmuller) [inline cli/cmd/user/add.go:51]: Would suggest phrasing it like this:
```
The user must be a member of %q to join one of its groups. Do you want to also invite the user to join %q?
```

File: cli/cmd/user/add.go [hunk 3]
```diff
@@ -37,16 +39,44 @@ func AddCmd(ch *cmdutil.Helper) *cobra.Command {
...
+						if err != nil {
+							return err
+						}
+
+						if addToOrg {
```
Comment (begelundmuller) [inline cli/cmd/user/add.go:56]: Again, can reduce nesting by handling the `if !addToOrg` case that exits early

File: cli/cmd/user/add.go [hunk 4]
```diff
@@ -37,16 +39,44 @@ func AddCmd(ch *cmdutil.Helper) *cobra.Command {
...
+							_, err = client.AddUsergroupMemberUser(cmd.Context(), &adminv1.AddUsergroupMemberUserRequest{
+								Organization:     ch.Org,
+								Usergroup:        group,
+								Email:            email,
+								AddRillUserToOrg: true,
```
Comment (begelundmuller) [inline cli/cmd/user/add.go:61]: You could also keep it simple by removing this flag, and just calling `AddOrganizationMemberUser` separately first. That would mean you don't need to handle the invite creation logic in multiple RPCs (and wouldn't need a `tx` in this RPC).

File: runtime/pkg/email/email.go [hunk 5]
```diff
@@ -211,12 +212,21 @@ func (c *Client) SendOrganizationInvite(opts *OrganizationInvite) error {
 	queryParams.Add("redirect", mustJoinURLPath(opts.FrontendURL, opts.OrgName))
 	finalURL := mustJoinURLPath(opts.AdminURL, "/auth/signup") + "?" + queryParams.Encode()
 
+	var body string
+	if opts.RoleName != "" {
+		body = fmt.Sprintf("%s has invited you to join <b>%s</b> as a %s for their Rill account. Get started interacting with fast, exploratory dashboards by clicking the button below to sign in and accept your invitation.", opts.InvitedByName, opts.OrgName, opts.RoleName)
+	} else if opts.UsergroupName != "" {
+		body = fmt.Sprintf("%s has invited you to join <b>%s</b> user group for their <b>%s</b> Rill account. Get started interacting with fast, exploratory dashboards by clicking the button below to sign in and accept your invitation.", opts.InvitedByName, opts.UsergroupName, opts.OrgName)
```
Comment (begelundmuller) [inline runtime/pkg/email/email.go:219]:
```suggestion
		body = fmt.Sprintf("%s has invited you to join the <b>%s</b> group for their <b>%s</b> Rill account. Get started interacting with fast, exploratory dashboards by clicking the button below to sign in and accept your invitation.", opts.InvitedByName, opts.UsergroupName, opts.OrgName)
```

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: admin/database/postgres/postgres.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/jackc/pgconn"
 	"github.com/jackc/pgtype"
 	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
```
Comment (begelundmuller) [inline admin/database/postgres/postgres.go:15]: This is the wrong Postgres library (we use `pgx`), so should be avoided

File: admin/database/postgres/postgres.go [hunk 7]
```diff
@@ -1950,6 +1973,20 @@ func (m *magicAuthTokenWithUserDTO) AsModel() (*database.MagicAuthTokenWithUser,
 	return m.MagicAuthTokenWithUser, nil
 }
 
+type OrganizationInviteDTO struct {
```
Comment (begelundmuller) [inline admin/database/postgres/postgres.go:1976]: nit: doesn't need to be exported, so should be `organizationInviteDTO`

File: admin/server/usergroups.go [hunk 8]
```diff
@@ -487,22 +487,41 @@ func (s *Server) AddUsergroupMemberUser(ctx context.Context, req *adminv1.AddUse
...
+		invite, err := s.admin.DB.FindOrganizationInvite(ctx, org.ID, req.Email)
+		if err != nil {
+			if !errors.Is(err, database.ErrNotFound) {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			// there is no pending invite return error
+			return nil, status.Error(codes.FailedPrecondition, "user is not a member of the organization")
+		}
+		// add group to the invite
+		invite.UsergroupIDs = append(invite.UsergroupIDs, group.ID)
```
Comment (begelundmuller) [inline admin/server/usergroups.go:510]: What if `group.ID` is already in `invite.UsergroupIDs`?

File: admin/user.go [hunk 9]
```diff
@@ -82,6 +82,25 @@ func (s *Service) CreateOrUpdateUser(ctx context.Context, email, name, photoURL
...
+
+			if !exists {
+				// ignore if usergroup does not exist, might have been deleted before invite was accepted
+				continue
+			}
+
+			err = s.DB.InsertUsergroupMemberUser(ctx, usergroupID, user.ID)
+			if err != nil {
+				return nil, err
+			}
```
Comment (begelundmuller) [inline admin/user.go:101]: Maybe for safety, gracefully handle if the user is already a member (I guess it would be a `database.ErrNotUnique`)? In case of duplicates in the list or retried calls

File: cli/cmd/user/add.go [hunk 10]
```diff
@@ -37,12 +37,54 @@ func AddCmd(ch *cmdutil.Helper) *cobra.Command {
...
 				})
 				if err != nil {
+					if strings.Contains(err.Error(), "user is not a member of the org") {
+						// ask if user should be added to the organization
+						ok, err := cmdutil.ConfirmPrompt(fmt.Sprintf("The user must be a member of %q to join one of its groups. Do you want to invite the user to join %q?", ch.Org, ch.Org), "", false)
+						if err != nil {
+							return err
+						}
+
+						if ok {
```
Comment (begelundmuller) [inline cli/cmd/user/add.go:53]: Can reduce nesting by handling the `if !ok` case that returns early

File: go.mod [hunk 11]
```diff
@@ -38,6 +38,7 @@ require (
 	github.com/fatih/color v1.15.0
 	github.com/fsnotify/fsnotify v1.7.0
 	github.com/getkin/kin-openapi v0.126.0
+	github.com/go-git/go-billy/v5 v5.5.0
```
Comment (begelundmuller) [inline go.mod:41]: Looks like an unrelated change to `go.mod`? Try running `go mod tidy`

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: admin/server/usergroups.go (hunk 8 above)
Comment (pjain1) [reply admin/server/usergroups.go:510]: oh yeah thought of that but forgot to dedupe

File: admin/user.go (hunk 9 above)
Comment (pjain1) [reply admin/user.go:101]: this is happening inside a transaction so in case of err tx fails, this needs to be checked explicitly. This is first time login case so If we deduplicate during adding the groups to invite, I think it should not be required to check ?

Comment (pjain1) [reply admin/user.go:101]: since this is happening inside transaction then in case of retries if usergroup was updated then invite would have been deleted as well so should not be a problem

File: admin/server/usergroups.go [hunk 12]
```diff
@@ -620,3 +641,12 @@ func memberUsergroupToPB(member *database.MemberUsergroup) *adminv1.MemberUsergr
...
 }
+
+func contains(slice []string, item string) bool {
+	for _, v := range slice {
+		if v == item {
+			return true
+		}
+	}
+	return false
+}
```
Comment (begelundmuller) [inline admin/server/usergroups.go:652]: You can use the standard library's `slices.Contains` function

File: admin/user.go (hunk 9 above)
Comment (begelundmuller) [reply admin/user.go:101]: Yeah I guess it's fine then :)

File: go.mod (hunk 11 above)
Comment (begelundmuller) [reply go.mod:41]: This line is still there? It's not referenced in this PR, so seems like a mistake?

**Review by begelundmuller — APPROVED** [review]
